### PR TITLE
[dbnode] Update the logic of index fileset volume cleanup

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Updates the logic of index fileset volume cleanup. This change mitigates the cases where recent volumes were not cleaned up after the shards owned by the dbnode changes.

For example, if we have following volumes:
```
volume index | shard list
0            | 0, 1
1            | 0, 1   // shard ownership changes due to modified topology
2            | 0
3            | 0
4            | 0
``` 

Previous cleanup logic would fail to delete some redundant volumes:
```
volume index | shard list
1            | 0, 1
2            | 0
3            | 0
4            | 0
```

After this change, the cleanup would leave following volumes
```
volume index | shard list
1            | 0, 1
4            | 0
```

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/dbnode/storage/index_test.go [hunk 1]
```diff
@@ -156,6 +156,150 @@ func TestNamespaceIndexCleanupDuplicateFilesets(t *testing.T) {
 	require.NoError(t, idx.CleanupDuplicateFileSets())
 }
 
+func TestNamespaceIndexCleanupDuplicateFilesets_GroupingByBlockStartAndVolume(t *testing.T) {
```
Comment (vpranckaitis) [inline src/dbnode/storage/index_test.go:159]: This test is unrelated to the change. I added it to improve coverage of grouping logic in the cleanup method.

File: src/dbnode/storage/index_test.go [hunk 2]
```diff
@@ -156,6 +156,150 @@ func TestNamespaceIndexCleanupDuplicateFilesets(t *testing.T) {
...
+		},
+		{
+			shards:       []uint32{1, 2},
+			shouldRemove: false,
+		},
+	}
+
+	blockStart := time.Now().Truncate(2 * time.Hour)
+	expectedFilesToRemove := make([]string, 0)
+	infoFiles := make([]fs.ReadIndexInfoFileResult, 0)
```
Comment (linasm) [inline src/dbnode/storage/index_test.go:274]: Nit: `var slice []type` is a bit cleaner syntax to declare an empty slice var.

File: src/dbnode/storage/index_test.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/storage/index_test.go:159]: Is it by `Volume`, or by `VolumeType`?

File: src/dbnode/storage/index_test.go [hunk 3]
```diff
@@ -612,6 +756,31 @@ func verifyFlushForShards(
 	require.Equal(t, expectedDocs, actualDocs)
 }
 
+func readIndexInfoFileResult(
```
Comment (linasm) [inline src/dbnode/storage/index_test.go:759]: Nit: please think of a better name, eg. `newReadIndexInfoFileResult` or smth.
The current name incorrectly implies that this function is reading something.

**Review by linasm — COMMENTED** [review]
Left a few comments.

File: src/dbnode/storage/index_test.go (hunk 1 above)
Comment (linasm) [reply src/dbnode/storage/index_test.go]: As I understand, they are grouped by block start and volume type (which the test name is supposed to reflect?), and within each such group they are ordered by volume index (which is perhaps just a cleanup logics implementation detail with no goal of being tested here)?
The essence being that cleanup decisions are being scoped to `BlockStartAndVolumeType`?

Comment (vpranckaitis) [reply src/dbnode/storage/index_test.go]: https://github.com/m3db/m3/pull/3409/commits/71ba88577cb6eea97786e4cef6a5f27b816b65bd

**Review by linasm — APPROVED** [review]
LGTM but you may want to wait for another pair of eyes on it (@notbdu perhaps).

File: src/dbnode/storage/index.go [hunk 4]
```diff
@@ -2137,20 +2137,18 @@ func (i *nsIndex) CleanupDuplicateFileSets() error {
...
-			currSegments := make([]fs.Segments, 0)
+			segmentsToKeep := make([]fs.Segments, 0)
 			for _, seg := range segmentsOrderByVolumeIndex {
-				if seg.ShardTimeRanges().IsSuperset(shardTimeRangesCovered) {
-					// Mark dupe segments for deletion.
-					for _, currSeg := range currSegments {
-						filesToDelete = append(filesToDelete, currSeg.AbsoluteFilePaths()...)
+				for len(segmentsToKeep) > 0 {
+					idx := len(segmentsToKeep) - 1
+					if previous := segmentsToKeep[idx]; seg.ShardTimeRanges().IsSuperset(previous.ShardTimeRanges()) {
```
Comment (notbdu) [inline src/dbnode/storage/index.go:2144]: Do we need to extract the active shards to compare here? 

Shouldn't `IsSuperset()` return `false` if the previous shard time ranges contain a shard missing from the current ranges? Like when we remove a shard.

```
// Check if other ranges has any shards we do not have.
for shard := range other.Iter() {
        if _, ok := r.Get(shard); !ok {
                return false
        }
}
```

Comment (vpranckaitis) [reply src/dbnode/storage/index.go:2144]: Not quite sure I get what you mean. Do you say that only shards from the most recent volume should be taken into account (ignoring any extra shards that might be there in older volumes)? E.g. if I have volumes with shards:
```
volume index: shard list
0: 0, 1, 2, 3
1: 0, 1, 2
2: 0, 1
```
I should only check shards {0, 1} and their time ranges when comparing volumes? And  shards {2, 3} should be skipped, since those only appear in obsolete volumes?

Comment (linasm) [reply src/dbnode/storage/index.go]: @notbdu could you review these changes?

**Review by notbdu — APPROVED** [review]
LGTM
